### PR TITLE
[JBPM-9465] Add support for using the public API to retrieve a node type

### PR DESCRIPTION
**Pull request by fjtirado**
Implementing getNodeType

**Depends on**

[droolsjbpm-knowledge:484](https://github.com/kiegroup/droolsjbpm-knowledge/pull/484)

**JIRA**: 

[JBPM-9465]( https://github.com/fjtirado/droolsjbpm-knowledge/pull/new/JBPM-9465)

### CODE REVIEW

**Comment by fjtirado** [comment]
jenkins retest this

**Comment by elguardian** [comment]
jenkins retest this

**Review by gmunozfe — COMMENTED** [review]
Shouldn't AsyncEventNode also include the type of node?
Missing one test for THROW_EVENT

**Comment by elguardian** [comment]
Jenkins retest this

File: jbpm-flow/src/test/java/org/jbpm/workflow/core/node/NodeTypeTest.java [hunk 1]
```diff
@@ -0,0 +1,61 @@
+package org.jbpm.workflow.core.node;
```
Comment (gmunozfe) [inline jbpm-flow/src/test/java/org/jbpm/workflow/core/node/NodeTypeTest.java:1]: Missing header

**Review by gmunozfe — CHANGES_REQUESTED** [review]
Missing header in new test class.

Also following compilation errors:
```
[ERROR] ../jbpm/jbpm-flow/src/test/java/org/jbpm/workflow/core/node/NodeInnerClassesTest.java org.jbpm.workflow.core.node.CompositeNode is abstract; cannot be instantiated
[ERROR] ../jbpm/jbpm-flow/src/test/java/org/jbpm/ruleflow/core/validation/RuleFlowProcessValidatorTest.java:[188,39] org.jbpm.workflow.core.node.CompositeNode is abstract; cannot be instantiated
[ERROR] ../jbpm/jbpm-flow/src/test/java/org/jbpm/process/EventTest.java:[754,39] org.jbpm.workflow.core.node.CompositeNode is abstract; cannot be instantiated
```

**Comment by fjtirado** [comment]
jenkins retest this

**Comment by gmunozfe** [comment]
These tests

- org.jbpm.process.CompensationTest.testCompensationEventSubProcessSpecific
- org.jbpm.process.CompensationTest.testNestedCompensationBoundaryEventGeneral
- org.jbpm.process.CompensationTest.testNestedCompensationEventSubProcessSpecific
- org.jbpm.process.CompensationTest.testCompensationEventSubProcessGeneral
- org.jbpm.process.CompensationTest.testCompensationBoundaryEventGeneral
- org.jbpm.process.CompensationTest.testNestedCompensationBoundaryEventSpecific
- org.jbpm.process.CompensationTest.testNestedCompensationEventSubProcessGeneral
- org.jbpm.process.CompensationTest.testCompensationBoundaryEventSpecific
- org.jbpm.process.EventSubProcessTest.testNestedEventSubProcess


 are failing with 
```
java.lang.IllegalArgumentException: wrong number of arguments
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.jbpm.process.test.NodeCreator.createNode(NodeCreator.java:41)
```

One of the possible fixes is this (in _createNode_ method of **NodeCreator** class):

`T result = this.constructor.newInstance(new Object[this.constructor.getParameterCount()]);`

**Comment by afalhambra** [comment]
Jenkins retest this

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/IntermediateThrowEventHandler.java [hunk 2]
```diff
@@ -52,7 +53,7 @@
 	public static final String LINK_TARGET = "target";
 
 	protected Node createNode(Attributes attrs) {
-		return new ActionNode();
+        return new ActionNode(NodeType.THROW_EVENT);
```
Comment (afalhambra) [inline jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/IntermediateThrowEventHandler.java:56]: Compilation error - this constructor has `protected` access only

File: jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/ScriptTaskHandler.java [hunk 3]
```diff
@@ -21,20 +21,21 @@
...
+import org.kie.api.definition.process.NodeType;
 import org.w3c.dom.Element;
 import org.xml.sax.Attributes;
 import org.xml.sax.SAXException;
 
 public class ScriptTaskHandler extends AbstractNodeHandler {
     
     protected Node createNode(Attributes attrs) {
-        ActionNode result = new ActionNode();
+        ActionNode result = new ActionNode(NodeType.SCRIPT_TASK);
```
Comment (afalhambra) [inline jbpm-bpmn2/src/main/java/org/jbpm/bpmn2/xml/ScriptTaskHandler.java:32]: Compilation error - this constructor has `protected` access only

**Review by afalhambra — CHANGES_REQUESTED** [review]
Very nice one @fjtirado - just small changes required for the access to a protected constructor

**Review by afalhambra — APPROVED** [review]
Just wait for CI to turn green :)

**Comment by fjtirado** [comment]
run fdb

**Review by gmunozfe — APPROVED** [review]
Looks good to me, fantastic work @fjtirado !

**Pull request merged**